### PR TITLE
make enarx-keep-sgx-shim static-pie without linker script

### PR DESCRIPTION
**Pull request by haraldh**
- move Layout to enarx_keep_sgx_shim::Layout
- add and call rcrt1.c:_dyn_reloc() to relocate the dynamic symbols

### CODE REVIEW

File: enarx-keep-sgx-shim/src/rcrt1.c [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+    rel_size = dyn[DT_RELASZ];
+
+    for (; rel_size; rel += 3, rel_size -= 3 * sizeof(size_t)) {
+        if (!IS_RELATIVE(rel[1], 0))
+            continue;
+        size_t *rel_addr = (void *) (base + rel[0]);
+        *rel_addr = base + rel[2];
+    }
+}
+
```
Comment (connorkuehl) [inline enarx-keep-sgx-shim/src/rcrt1.c:55]: If this is identical to `enarx-keep-sev-shim/src/rcrt1.c`, can we put it in the repo root and place symbolic links in `enarx-keep-sev-shim/src/` and `enarx-keep-sgx-shim/src/` that point to it?

File: enarx-keep-sgx-shim/link.json [hunk 2]
```diff
@@ -1,12 +1,12 @@
...
         "prepend": [
-            "-Wl,--script=layout.ld",
-            "-Wl,--gc-keep-exported"
+            "-static-pie"
         ],
 
         "replace": {
-            ".*crt.\\.o": []
+            ".*/crt.\\.o": [],
+            "-no-pie": []
```
[the remark this answers is not in the document]
Comment (npmccallum) [reply enarx-keep-sgx-shim/link.json:9]: This is **removing** `-no-pie` from the compiler's command line.

File: enarx-keep-sgx-shim/src/rcrt1.c (hunk 1 above)
Comment (npmccallum) [reply enarx-keep-sgx-shim/src/rcrt1.c:55]: Alternatively, could we somehow make this a crate we can share?

File: enarx-keep-sgx-shim/link.json (hunk 2 above)
[the remark this answers is not in the document]
Comment (lkatalin) [reply enarx-keep-sgx-shim/link.json:9]: Oh, I see, I read it incorrectly but now it makes sense.

File: enarx-keep-sgx-shim/src/heap.rs [hunk 3]
```diff
@@ -19,7 +19,7 @@ pub struct Heap {
 }
 
 impl Heap {
-    pub unsafe fn new(heap: Span<u64>) -> Self {
+    pub unsafe fn new(heap: Span<usize>) -> Self {
```
[the remark this answers is not in the document]
Comment (haraldh) [reply enarx-keep-sgx-shim/src/heap.rs:22]: The original two versions of Layou differed ..  one had usize, one had u64

File: enarx-keep-sgx-shim/src/rcrt1.c (hunk 1 above)
Comment (haraldh) [reply enarx-keep-sgx-shim/src/rcrt1.c:55]: maybe... could be a follow up patch

File: enarx-keep-sgx-shim/src/entry.rs [hunk 4]
```diff
@@ -28,7 +28,7 @@ fn crt0setup<'a>(
     crt0: &'a mut [u8],
 ) -> Result<Handle<'a>, OutOfSpace> {
     let rand = unsafe { core::mem::transmute([random(), random()]) };
-    let phdr = layout.code.start + hdr.e_phoff;
+    let phdr = layout.code.start as u64 + hdr.e_phoff;
```
[the remark this answers is not in the document]
Comment (haraldh) [reply enarx-keep-sgx-shim/src/entry.rs:31]: The original two versions of Layou differed .. one had usize, one had u64

File: enarx-keep-sgx-shim/src/rcrt1.c [hunk 5]
```diff
@@ -0,0 +1,55 @@
+// SPDX-License-Identifier: MIT
+
+// Taken from rcrt1.c musl libc and reduced to x86_64
+// with custom arguments.
```
[the remark this answers is not in the document]
Comment (haraldh) [reply enarx-keep-sgx-shim/src/rcrt1.c:4]: https://git.musl-libc.org/cgit/musl/tree/ldso/dlstart.c

File: enarx-keep-sgx-shim/src/start.S [hunk 6]
```diff
@@ -107,7 +110,37 @@ enclave:
...
-    call    entry                                   # Jump to Rust
+
+    pushq   %rdi
+    pushq   %rsi
+    pushq   %rdx
+    pushq   %rcx
+    pushq   %r8
+    pushq   %r9
+    pushq   %r10
+    pushq   %r11
```
[the remark this answers is not in the document]
Comment (haraldh) [reply enarx-keep-sgx-shim/src/start.S:121]: If it is used more than once... yes

File: enarx-keep-sgx-shim/src/start.S [hunk 7]
```diff
@@ -183,7 +216,7 @@ enclave:
 syscall:
     savep                                           # Save preserved registers
     mov     %rsp,                   SRSP(%rcx)      # Save restoration stack pointer
-    
+
```
[the remark this answers is not in the document]
Comment (haraldh) [reply enarx-keep-sgx-shim/src/start.S:219]: typo of removing spaces on an empty line?

[the remark this answers is not in the document]
Comment (lkatalin) [reply enarx-keep-sgx-shim/src/start.S:219]: Oh, I see. More like a fix up then.

File: enarx-keep-sgx-shim/src/entry.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (lkatalin) [reply enarx-keep-sgx-shim/src/entry.rs:31]: Got it! I remember now. Thanks.

File: enarx-keep-sgx-shim/src/start.S (hunk 7 above)
[the remark this answers is not in the document]
Comment (haraldh) [reply enarx-keep-sgx-shim/src/start.S:219]: well my editor did this automatically and I just left it in the patch...

File: enarx-keep-sgx-shim/src/rcrt1.c (hunk 1 above)
Comment (haraldh) [reply enarx-keep-sgx-shim/src/rcrt1.c]: refactored in a separate crate and converted to rust :smile:

**Comment by lkatalin** [comment]
@haraldh Looks like it needs a rebase.

**Comment by haraldh** [comment]
forgot to add "rcrt1" to `Makefile.toml`

**Pull request merged**